kubernetes/kubernetes.go

```diff
@@ -545,57 +556,67 @@ func (ka KubeAgentConfig) sendMetricsToCustomS3(metricSample *os.File) {
 	}
 }
 
-func (ka KubeAgentConfig) sendMetricsToCustomBlob(metricSample *os.File) {
+func (ka KubeAgentConfig) sendMetricsToCustomBlob(ctx context.Context, metricSample *os.File) {
 	cred, err := azidentity.NewClientSecretCredential(ka.CustomAzureTenantID, ka.CustomAzureClientID,
 		ka.CustomAzureClientSecret, nil)
 	if err != nil {
 		log.Fatalf("Could not establish Azure credentials, "+
 			"ensure Azure environment variables are set correctly: %s", err)
 	}
-	azureClient, err := azblob.NewClient(ka.CustomAzureBlobURL, cred, nil)
+	retryConfig := azblob.ClientOptions{
+		ClientOptions: azcore.ClientOptions{
+			Retry: policy.RetryOptions{
+				MaxRetries: 3,
+			},
+		},
+	}
+	azureClient, err := azblob.NewClient(ka.CustomAzureBlobURL, cred, &retryConfig)
 	if err != nil {
 		log.Fatalf("Could not establish Azure Client, "+
 			"ensure Azure environment variables are set correctly: %s", err)
 	}
-	ka.uploadBlob(azureClient, metricSample)
+	ka.uploadBlob(ctx, azureClient, metricSample)
 }
 
-func (ka KubeAgentConfig) uploadBlob(client *azblob.Client, metricSample *os.File) {
+func (ka KubeAgentConfig) uploadBlob(ctx context.Context, client *azblob.Client, metricSample *os.File) {
 	file, err := os.Open(metricSample.Name())
 	if err != nil {
 		log.Fatalf("Unable to open metric sample file %v", err)
 	}
 	defer file.Close()
-	key := generateSampleKey(ka.clusterUID, azure)
-	_, err = client.UploadFile(context.Background(), ka.CustomAzureUploadBlobContainerName, key, file, nil)
+	key := generateSampleKey(ka.clusterUID, azureProvider)
+	_, err = client.UploadFile(ctx, ka.CustomAzureUploadBlobContainerName, key, file, nil)
 	if err != nil {
 		log.Fatalf("Failed to put Object to custom Azure blob with error: %s", err)
-	} else {
-		sn := strings.Split(metricSample.Name(), "/")
-		log.Infof("Exported metric sample %s to custom Azure blob: %s",
-			strings.TrimSuffix(sn[len(sn)-1], ".tgz"), ka.CustomAzureUploadBlobContainerName)
-	}
-	err = os.Remove(metricSample.Name())
-	if err != nil {
-		log.Warnf("Warning: Unable to cleanup after metric sample upload: %v", err)
 	}
+
+	sn := strings.Split(metricSample.Name(), "/")
+	log.Infof("Exported metric sample %s to custom Azure blob: %s",
+		strings.TrimSuffix(sn[len(sn)-1], ".tgz"), ka.CustomAzureUploadBlobContainerName)
+
+	defer func(name string) {
```

Comment:
Defers are operated on as a LIFO stack. In this case it will still remove before closing.

Reply:
Added a defer function and both close and remove are now handled LIFO within that function.